src/library/Cardano/CoinSelection.hs

```diff
@@ -11,38 +13,149 @@
 --
 module Cardano.CoinSelection
     (
-      -- * Types
-      CoinSelection (..)
+      -- * Coin
+      Coin (..)
+    , coinIsValid
+
+      -- * Coin Map
+    , CoinMap (..)
+    , CoinMapEntry (..)
+    , coinMapFromList
+    , coinMapToList
+    , coinMapValue
+    , coinMapRandomEntry
+
+      -- * Coin Selection
+    , CoinSelection (..)
+    , sumInputs
+    , sumOutputs
+    , sumChange
+
+      -- * Coin Selection Algorithm
     , CoinSelectionAlgorithm (..)
     , CoinSelectionOptions (..)
     , CoinSelectionError (..)
-    , Input (..)
-    , Output (..)
-
-      -- * Calculating Balances
-    , inputBalance
-    , outputBalance
-    , changeBalance
-    , feeBalance
 
     ) where
 
 import Prelude
 
-import Cardano.Types
-    ( Coin (..), UTxO (..) )
+import Control.Arrow
+    ( (&&&) )
+import Control.DeepSeq
+    ( NFData (..) )
 import Control.Monad.Trans.Except
     ( ExceptT (..) )
-import Data.List
-    ( foldl' )
-import Data.List.NonEmpty
-    ( NonEmpty (..) )
+import Crypto.Number.Generate
+    ( generateBetween )
+import Crypto.Random.Types
+    ( MonadRandom )
+import Data.Map.Strict
+    ( Map )
 import Data.Word
     ( Word64, Word8 )
 import Fmt
-    ( Buildable (..), blockListF, blockListF', listF, nameF )
+    ( Buildable (..), blockListF, listF, nameF )
 import GHC.Generics
     ( Generic )
+import Quiet
+    ( Quiet (Quiet) )
+
+import qualified Data.Foldable as F
+import qualified Data.Map.Strict as Map
+
+--------------------------------------------------------------------------------
+-- Coin
+--------------------------------------------------------------------------------
+
+-- | A non-negative integer value that represents a number of Lovelace.
+--
+-- One Ada is equal to 1,000,000 Lovelace.
+--
+newtype Coin = Coin
+    { unCoin :: Word64 }
+    deriving stock (Eq, Generic, Ord)
+    deriving Show via (Quiet Coin)
+
+instance Monoid Coin where
+    mempty = minBound
+
+instance Semigroup Coin where
+    Coin a <> Coin b = Coin (a + b)
+
+instance NFData Coin
+
+instance Bounded Coin where
+    minBound =
+        Coin 0
+    maxBound =
+        Coin 45_000_000_000_000_000
```

Comment:
I always was wondering if this would not be good to have wrapper type on top of `Coin` that encodes ADA/LOVELACE information. Something like Quantity "ada"/Quantity "lovelace" that enables to talk with whatever "level" and always be correct....

Reply:
> I always was wondering if this would not be good to have wrapper type on top of `Coin` that encodes ADA/LOVELACE information. Something like Quantity "ada"/Quantity "lovelace" that enables to talk with whatever "level" and always be correct....

Agreed, that would definitely be nice to have! Though luckily, we don't need something too complicated for this library, as:

1. We only have one coin type, which encodes a number of Lovelace.
2. We never have to deal with values that are smaller than 1 Lovelace.